nuxeo-runtime/nuxeo-stream/src/main/java/org/nuxeo/lib/stream/log/kafka/KafkaLogTailer.java

```diff
@@ -0,0 +1,406 @@
+        return ret;
+    }
+
+    public static <M extends Externalizable> KafkaLogTailer<M> createAndSubscribe(String prefix, Collection<String> names, String group, Properties consumerProps,
+                                                                                  RebalanceListener listener) {
+        KafkaLogTailer<M> ret = new KafkaLogTailer<>(prefix, group, consumerProps);
+        ret.id = buildSubscribeId(ret.group, names);
+        ret.names = names;
+        Collection<String> topics = names.stream().map(name -> prefix + name).collect(Collectors.toList());
+        ret.listener = listener;
+        ret.consumer.subscribe(topics, ret);
+        ret.partitions = Collections.emptyList();
+        ret.topicPartitions = Collections.emptyList();
+        log.debug(String.format("Created tailer with subscription: %s using prefix: %s", ret.id, prefix));
+        return ret;
+    }
+
+    protected static String buildId(String group, Collection<LogPartition> partitions) {
+        return group + ":" + partitions.stream().map(LogPartition::toString).collect(Collectors.joining("|"));
+    }
+
+    protected static String buildSubscribeId(String group, Collection<String> names) {
+        return group + ":" + names.stream().collect(Collectors.joining("|"));
+    }
+
+
+    @Override
+    public LogRecord<M> read(Duration timeout) throws InterruptedException {
+        if (closed) {
+            throw new IllegalStateException("The tailer has been closed.");
+        }
+        if (records.isEmpty()) {
+            int items = poll(timeout);
+            if (isRebalanced) {
+                isRebalanced = false;
+                log.debug("Rebalance happens during poll, raising exception");
+                throw new RebalanceException("Partitions has been rebalanced");
+            }
+            if (items == 0) {
+                if (log.isTraceEnabled()) {
+                    log.trace("No data " + id + " after " + timeout.toMillis() + " ms");
+                }
+                return null;
+            }
+        }
+        ConsumerRecord<String, Bytes> record = records.poll();
+        lastOffsets.put(new TopicPartition(record.topic(), record.partition()), record.offset());
+        M value = messageOf(record.value());
+        LogPartition partition = LogPartition.of(getNameForTopic(record.topic()), record.partition());
+        LogOffset offset = new LogOffsetImpl(partition, record.offset());
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("Read from %s/%s, key: %s, value: %s", offset, group, record.key(), value));
+        }
+        return new LogRecord<>(value, offset);
+    }
+
+    protected String getNameForTopic(String topic) {
+        return topic.replaceFirst(prefix, "");
+    }
+
+    @SuppressWarnings("unchecked")
+    protected M messageOf(Bytes value) {
+        ByteArrayInputStream bis = new ByteArrayInputStream(value.get());
+        ObjectInput in = null;
+        try {
```

Comment:
Curious to know why try-auto-close couldn't not be used here ?

Reply:
don't remember :)

Reply:
:)